src/db/file.rs

```diff
@@ -182,64 +193,40 @@ pub fn add_path_into_database<P: AsRef<Path>>(conn: &Connection,
-                    match s3_res {
-                        // we've successfully uploaded the content, so steal it;
-                        // we don't want to put it in the DB
-                        Ok(_) => break None,
-                        // Since s3 was configured, we want to panic on failure to upload.
-                        Err(e) => {
-                            log::error!("failed to upload to {}: {:?}", bucket_path, e);
-                            // Get a new client, in case the old one's connection is stale.
-                            // AWS will kill our connection if it's alive for too long; this avoids
-                            // that preventing us from building the crate entirely.
-                            *client = s3_client().unwrap();
-                            if attempts > 3 {
-                                panic!("failed to upload 3 times, exiting");
-                            } else {
-                                continue;
-                            }
-                        },
-                    }
-                }
+            if let Some(client) = &client {
+                futures.push(client.put_object(PutObjectRequest {
+                    bucket: "rust-docs-rs".into(),
+                    key: bucket_path.clone(),
+                    body: Some(content.clone().into()),
+                    content_type: Some(mime.clone()),
+                    ..Default::default()
+                }));
             } else {
-                Some(content.clone().into())
-            };
+                // If AWS credentials are configured, don't insert/update the database
+                // check if file already exists in database
+                let rows = try!(conn.query("SELECT COUNT(*) FROM files WHERE path = $1", &[&bucket_path]));
 
-            file_list_with_mimes.push((mime.clone(), file_path.clone()));
-
-            (
-                bucket_path,
-                content,
-                mime,
-            )
-        };
-
-        // If AWS credentials are configured, don't insert/update the database
-        if client.is_none() {
-            // check if file already exists in database
-            let rows = try!(conn.query("SELECT COUNT(*) FROM files WHERE path = $1", &[&path]));
+                if rows.get(0).get::<usize, i64>(0) == 0 {
+                    try!(trans.query("INSERT INTO files (path, mime, content) VALUES ($1, $2, $3)",
+                                    &[&bucket_path, &mime, &content]));
+                } else {
+                    try!(trans.query("UPDATE files SET mime = $2, content = $3, date_updated = NOW() \
+                                    WHERE path = $1",
+                                    &[&bucket_path, &mime, &content]));
+                }
+            }
 
-            let content = content.expect("content never None if client is None");
+            file_list_with_mimes.push((mime.clone(), file_path.clone()));
```

Comment:
I think if this happens during a retry, it will duplicate items in the `file_list_with_mimes` vec.

This might need to be done only when the `put_object` is successful?

Something like

```
client.put_object(PutObjectRequest {
  // ...
}).and_then(|| {
  file_list_with_mimes.push((mime.clone(), file_path.clone()));
});
```

except I think now `file_list_with_mimes` needs to be in an `Arc`

Reply:
Might be simpler to deduplicate after uploading by file path.

Reply:
Or prevent duplicates during insert?